compiler/rustc_infer/src/infer/combine.rs

```diff
@@ -535,13 +539,17 @@ impl TypeRelation<'tcx> for Generalizer<'_, 'tcx> {
     fn tys(&mut self, t: Ty<'tcx>, t2: Ty<'tcx>) -> RelateResult<'tcx, Ty<'tcx>> {
         assert_eq!(t, t2); // we are abusing TypeRelation here; both LHS and RHS ought to be ==
 
+        let cache_key = (t, t2);
```

Comment:
Actually there's no need to include both t and t2 in the key since they're both equal! (see above assertion)

This could help performance even more..

Reply:
Fixed in latest commit